File: backtracking/balancing.java
// class AllBraces{
//   static void print(ArrayList<ArrayList<Character>> arr) {
//     for (int i = 0; i < arr.size(); i++) {
//       System.out.println(arr.get(i).toString());
//     }
//   }

//   static void printAllBacesRec(
//     int n,
//     int leftCount,
//     int rightCount, ArrayList<Character> output,
//     ArrayList<ArrayList<Character>> result) {

//     if (leftCount >= n && rightCount >=n) {
//       result.add((ArrayList)output.clone());
//     }

//     if (leftCount < n) {
//       output.add('{');
//       printAllBacesRec(n, leftCount + 1, rightCount, output, result);
//       output.remove(output.size() - 1);
//     }

//     if (rightCount < leftCount) {
//       output.add('}');
//       printAllBacesRec(n, leftCount, rightCount + 1, output, result);
//       output.remove(output.size() - 1);
//     }
//   }

//   static ArrayList<ArrayList<Character>> printAllBraces(int n) {
//     ArrayList<ArrayList<Character>> result = new ArrayList<ArrayList<Character>>();
//     ArrayList<Character> output = new ArrayList<Character>();
//     printAllBacesRec(n, 0, 0, output, result);
//     return result;
//   }
  
//   public static void main(String[] args) {
//     ArrayList<ArrayList<Character>> result = new ArrayList<ArrayList<Character>>();
//     result =  printAllBraces(3);
//     print (result);
//   }
// }  

import java.util.*;

class AllBraces {
    static void print(ArrayList<ArrayList<Character>> arr) {
        for (int i=0; i < arr.size(); i++) {
            System.out.println(arr.get(i).toString());
        }
    }

    static void printAllBracesRec(
        int n,
        int leftCount,
        int rightCount, ArrayList<Character> output,
        ArrayList<ArrayList<Character>> result) {
            if (leftCount >= n && rightCount >= n) {
                result.add((ArrayList)output.clone());
            }
            
            if(leftCount < n) {
                output.add('{');
                printAllBracesRec(n, leftCount + 1, rightCount, output, result);
                output.remove(output.size() - 1);
            }

            if (rightCount < leftCount) {
                output.add('}');
                printAllBracesRec(n, leftCount, rightCount + 1, output, result);
                output.remove(output.size() - 1);
            }
        }

    static ArrayList<ArrayList<Character>> printAllBraces(int n) {
        ArrayList<ArrayList<Character>> result = new ArrayList<ArrayList<Character>>();
        ArrayList<Character> output = new ArrayList<Character>();
        printAllBracesRec(n, 0, 0, output, result);
        return result;
    }

    public static void main(String[] args) {
        ArrayList<ArrayList<Character>> result = new ArrayList<ArrayList<Character>>();
        result = printAllBraces(3);
        print(result);
    }

}

File: backtracking/balancing.cpp
// void print(vector<vector<char>> result){
//   for(int i = 0; i < result.size(); i++){
//     cout << "[ ";
//     for(int j = 0; j < result[i].size(); j++){
//       cout << result[i][j] << ", "; 
//     }
//     cout << "]" << endl;
//   }
// }

// void print_all_braces_rec(
//     int n,
//     int left_count,
//     int right_count,
//     vector<char>& output, vector<vector<char>>& result) {

//   if (left_count == n && right_count == n) {
//     result.push_back(output);
//   }

//   if (left_count < n) {
//     output.push_back('{');
//     print_all_braces_rec(n, left_count + 1, right_count, output, result);
//     output.pop_back();
//   }

//   if (right_count < left_count) {
//     output.push_back('}');
//     print_all_braces_rec(n, left_count, right_count + 1, output, result);
//     output.pop_back();
//   }
// }

// vector<vector<char>> print_all_braces(int n) {
//   vector<vector<char>> result;
//   vector<char> output;
//   print_all_braces_rec(n, 0, 0, output, result);
//   return result;
// }

// int main() {
//   vector<vector<char>> result = print_all_braces(3);
//   print(result);
// }


voud print(vector<vector<char>> result) {
    for (int i = 0; i < result.size(); i++) {
        cout << "[";
        for(int j = 0; j < result[i].size(); j++) {
            cout << result[i][j] << ", ";
        }
        cout << "]" << endl;
    }
}

void print_all_braces_rec(
    int n,
    int left_count,
    int right_count,
    vector<char>& output, vector<vector<char>>& result) {
        if (left_count == n && right_count == n) {
            result.push_back(output);
        }
        if (left_count < n) {
            output.push_back('{');
            print_all_braces_rec(n, left_count + 1, right_count, output, result);
            output.pop_back();
        }
        if (right_count < left_count) {
            output.push_back('}');
            print_all_braces_rec(n, left_count, right_count + 1, output, result);
            output.pop_back();
        }
    }

    vector<vector<char>> print_all_braces(int n) {
        vector<vector<char>> result;
        vector<char> output;
        print_all_braces_rec(n, 0, 0, output, result);
        return result;
    }

    int main() {
        vector<vector<char>> result = print_all_braces(3);
        print(result);
    }